### PR TITLE
Filtration by `dom_consistency` while building small step semantics

### PR DESCRIPTION
**Pull request by volodeyka**
Here in definition of state configuration:
https://github.com/volodeyka/event-struct/blob/6796a493653fc4149f77a3a6efe146303ec669db/regmachine.v#L90
we can have not only `dom_consistent` `exec_event_struct`'es (i.e with irreflexive `cf`). Also in `eval_step` we are obtaining not only  `dom_consistent` `exec_event_struct`es. In this PR I tried to deal with these problems by filtering all possible `exec_event_struct`es  while building small step semantics by some predicate from witch we can obtain `dom_consistency`.

---
This change is [Reviewable](https://reviewable.io/reviews/volodeyka/event-struct/66)

### CODE REVIEW

File: eventstructure.v [hunk 1]
```diff
@@ -135,14 +135,18 @@ Structure fin_exec_event_struct := Pack {
   dom_sorted : sorted (>%O) dom;
   lprf       : { fsfun for fun e =>
                    {| lab_prj := ThreadEnd; fpred_prj := e; frf_prj := e |} };
-  _          : {subset (finsupp lprf) <= dom};
+  _          : all (fun x => x \in dom) (finsupp lprf);
```
Comment (anton-trunov) [inline eventstructure.v:138]: I haven't reviewed this PR properly yet but here is one quick remark:
is it possible to use `finsupp lprf \subset dom` here? (the  `_ \subset _` relation of the `fintype` library)?

Admittedly, this would force us to revisit the design here which we might not want to do due to time constraints.

**Comment by anton-trunov** [comment]
Looks like you need Mathcomp of version at least 1.12.0 to use that contraposition lemma.

File: eventstructure.v (hunk 1 above)
Comment (eupp) [reply eventstructure.v:138]: Why do you prefer to use `fintype` theory here?

**Review by eupp — COMMENTED** [review]
Could you please elaborate more on what is exactly done here? 
Now I can see at least two orthogonal parts of this PR:
1) an `eqType` instance for `fin_exec_event_struct` (and related changes, i.e. usage of `all`)
2) some new definitions to build consistent event structures for regmachine semantics. 

Unfortunately, it's still hard to follow the code in `transtitionsystem.v` and `regmachine.v`. 
I have a few suggestions on how to improve the readability. 
First of all, it would be useful to add type-annotations to all definitions. 
The types would self-document the code and simplify reading. 
I have other ideas, but it would be more convenient to discuss them in a separate PR. 

So, I propose to split this at least into two PRs:
1) `eqType` instance for `fin_exec_eventstruct`. There we can discuss what design is better (using `all` or `subset` from `fintype`). 
2) new functionality related to consistent event structures.

File: eventstructure.v [hunk 2]
```diff
@@ -430,51 +430,68 @@ Proof. by move=> I; apply/cfP; exists e1, e2; rewrite !ca_refl I. Qed.
 
 Notation cf_step e1 e2 :=
   [|| icf e1 e2,
-      has (cf e1) (fica e2) |
-      has (cf e2) (fica e1)].
+      has (cf e1) (filter (fun x => e2 != x) (fica e2)) |
+      has (cf e2) (filter (fun x => e1 != x) (fica e1))].
```
Comment (anton-trunov) [inline eventstructure.v:434]: We should document the intended semantics of `fica`. My gut feeling tells me it shouldn't allow `e` to be its own immediate cause. And this change kind of backs this feeling.

File: regmachine.v [hunk 3]
```diff
@@ -30,6 +30,9 @@ From event_struct Require Import transitionsystem ident.
 (*        es_seq == takes event structure `es`, location `x`, predsessor event*)
 (*          `pr` and returns sequence of `es + Read x v`, where v runs on all *)
 (*           values  `v` that we can read in location `x`                     *)
+(*   ces_seq_aux == all dom_consistent exec_eventstructures from es_seq       *)
+(*       ces_seq == ces_seq_aux mapped by Consist (doing so we are otaining   *)
+(*           cexec_eventstructures form consistent exec_eventstructures)      *)
```
Comment (anton-trunov) [inline regmachine.v:35]: I'd suggest rewording this more informally and provide some intuition here. These comments should not just recapitulate the code, in my opinion.

File: eventstructure.v (hunk 1 above)
Comment (volodeyka) [reply eventstructure.v:138]: > is it possible to use `finsupp lprf \subset dom` here? (the `_ \subset _` relation of the `fintype` library)?

It seems like to use `finsupp lprf \subset dom` `E` needs to be `finType`, because `dom : seq E`: I have got an error saying something like "`dom` has type `mem_pred E` while it is expected to have type `mem_pred ?T`". Maybe I should import something to fix this?

File: eventstructure.v (hunk 2 above)
Comment (volodeyka) [reply eventstructure.v:434]: Now, I don't remember exactly why `fica` allows `e` to be its own immediate cause, such definition makes sense when it comes to proving certain facts of the "transitionsystem". I don't know whether it is very useful now, but due to the lack of time I think we can come back to this question later.

File: regmachine.v [hunk 4]
```diff
@@ -1,5 +1,5 @@
 From mathcomp Require Import ssreflect ssrfun ssrbool ssrnat seq.
-From mathcomp Require Import eqtype choice finfun finmap.
+From mathcomp Require Import eqtype choice finfun finmap tuple.
```
Comment (anton-trunov) [inline regmachine.v:2]:
```suggestion
From mathcomp Require Import eqtype choice finfun finmap.
```

File: eventstructure.v (hunk 2 above)
Comment (eupp) [reply eventstructure.v:459]: I think this is because in `transitionsystem.v` we have a transition relation annotated by the new event and its parent (in program order). Then, in order to get rid of `option`, we always pass some parent event, but for the roots, we pass the event itself. Once we'll generalize to a uniform representation of causality through one function `covering : E -> seq E` we'll be able to resolve this issue and pass an empty list of parents for the root events.

File: regmachine.v (hunk 3 above)
Comment (eupp) [reply regmachine.v]: I agree, the current description doesn't help much. 
But I think you can fix it later in a separate PR.

**Review by eupp — COMMENTED** [review]
Once the CI is fixed, I'm okay with merging this.
